nautobot/extras/models/metadata.py

```diff
@@ -135,3 +138,108 @@ def save(self, *args, **kwargs):
+            max_length=CHARFIELD_MAX_LENGTH,
+        ),
+        help_text="List of scoped fields, only direct fields on the model",
+    )
+    value = models.JSONField(
+        blank=True,
+        null=True,
+        help_text="Relevant data value to an object field or a set of object fields",
+    )
+    assigned_object_type = models.ForeignKey(to=ContentType, on_delete=models.SET_NULL, null=True, related_name="+")
+    assigned_object_id = models.UUIDField(db_index=True)
+    assigned_object = GenericForeignKey(ct_field="assigned_object_type", fk_field="assigned_object_id")
+
+    natural_key_field_names = ["pk"]
+    documentation_static_path = "docs/user-guide/platform-functionality/metadata.html"
+
+    class Meta:
+        ordering = ["metadata_type"]
+
+        indexes = [
+            models.Index(
+                name="assigned_object",
+                fields=["assigned_object_type", "assigned_object_id"],
+            ),
+            models.Index(
+                name="assigned_object_contact",
+                fields=["assigned_object_type", "assigned_object_id", "contact"],
+            ),
+            models.Index(
+                name="assigned_object_team",
+                fields=["assigned_object_type", "assigned_object_id", "team"],
+            ),
+        ]
+
+    def __str__(self):
+        if self.contact:
+            return f"{self.contact}: {self.metadata_type} - {self.assigned_object} - {self.value}"
+        else:
+            return f"{self.team}: {self.metadata_type} - {self.assigned_object} - {self.value}"
+
+    def clean(self):
+        super().clean()
+        if self.contact is None and self.team is None:
+            raise ValidationError("Either a contact or a team must be specified")
+        if self.contact is not None and self.team is not None:
+            raise ValidationError("A contact and a team cannot be both specified at once")
+
+        if self.present_in_database:
+            # Check immutable fields
+            database_object = self.__class__.objects.get(pk=self.pk)
+            if self.metadata_type != database_object.metadata_type:
+                raise ValidationError({"metadata_type": "Cannot be changed once created"})
+
+        # Check if there is any intersections of scoped_fields of ObjectMetadata instances in the database.
+        object_metadata_scoped_fields = (
+            ObjectMetadata.objects.filter(
+                metadata_type=self.metadata_type,
+                assigned_object_type=self.assigned_object_type,
+                assigned_object_id=self.assigned_object_id,
+            )
+            .exclude(pk=self.pk)
+            .values_list("scoped_fields", flat=True)
+        )
+        duplicate_scoped_fields_list = set([])
+
+        for scoped_fields in object_metadata_scoped_fields:
+            for field in scoped_fields:
+                if field in self.scoped_fields:
+                    duplicate_scoped_fields_list.add(field)
```

Comment:
```suggestion
            duplicate_scoped_fields_list |= set(self.scoped_fields).intersection(scoped_fields)
```

Or something like that at least.